experimental/dds/tree2/src/feature-libraries/modular-schema/crossFieldQueries.ts

```diff
@@ -5,11 +5,106 @@
 
 import { assert } from "@fluidframework/common-utils";
 import { RevisionTag } from "../../core";
-import { brand, NestedSet } from "../../util";
+import { brand, getOrAddInMap } from "../../util";
 import { IdAllocator } from "./fieldChangeHandler";
 import { ChangesetLocalId } from "./modularChangeTypes";
 
-export type CrossFieldQuerySet = NestedSet<RevisionTag | undefined, ChangesetLocalId>;
+export type CrossFieldMap<T> = Map<RevisionTag | undefined, IdRangeMap<T>>;
+export type CrossFieldQuerySet = CrossFieldMap<boolean>;
+
+export function addCrossFieldQuery(
+	set: CrossFieldQuerySet,
+	revision: RevisionTag | undefined,
+	id: ChangesetLocalId,
+	count: number,
+): void {
+	const rangeMap = getOrAddInMap(set, revision, []);
+	setInRangeMap(rangeMap, id, count, true);
+}
+
+export function getFirstFromRangeMap<T>(
+	map: IdRangeMap<T>,
+	id: ChangesetLocalId,
+	count: number,
+): CrossFieldRange<T> | undefined {
+	for (const range of map) {
+		if (range.id >= (id as number) + count) {
+			break;
+		}
+
+		if ((range.id as number) + range.length > id) {
+			return range;
+		}
+	}
+
+	return undefined;
+}
+
+export function setInCrossFieldMap<T>(
+	map: CrossFieldMap<T>,
+	revision: RevisionTag | undefined,
+	id: ChangesetLocalId,
+	count: number,
+	value: T,
+): void {
+	setInRangeMap(getOrAddInMap(map, revision, []), id, count, value);
+}
+
+export function getFirstFromCrossFieldMap<T>(
+	map: CrossFieldMap<T>,
+	revision: RevisionTag | undefined,
+	id: ChangesetLocalId,
+	count: number,
+): CrossFieldRange<T> | undefined {
+	return getFirstFromRangeMap(map.get(revision) ?? [], id, count);
+}
+
+export function setInRangeMap<T>(
+	map: IdRangeMap<T>,
+	id: ChangesetLocalId,
+	count: number,
+	value: T,
+): void {
+	const newEntry: CrossFieldRange<T> = { id, length: count, data: value };
+	for (const [i, range] of map.entries()) {
+		const lastRangeId = (range.id as number) + range.length - 1;
+		const lastTargetId = (id as number) + count - 1;
+
+		if (range.id > lastTargetId) {
+			break;
+		}
+
+		if (lastRangeId < id) {
+			continue;
+		}
+
+		if (range.id < id && lastRangeId > lastTargetId) {
+			// range has excess portions both before and after target range
+			// We replace the existing range with a range before and a range after and insert the new range
+			range.length = (id as number) + count - range.id;
+			map.splice(i + 1, 0, newEntry);
+			map.splice(i + 2, 0, {
+				id: brand((id as number) + 1),
+				length: lastRangeId - id,
+				data: range.data,
+			});
+			return;
+		}
+
+		if (range.id < id) {
+			range.length = id - range.id;
```

Comment:
Add comment to explain that it's okay not to preserve the latter part of this range because it will be fully overwritten